crates/aiken-lang/src/tipo/environment.rs

```diff
@@ -1415,10 +1415,15 @@ impl<'a> Environment<'a> {
             && !(t1.is_function() || t2.is_function())
             && !(t1.is_generic() || t2.is_generic())
             && !(t1.is_string() || t2.is_string())
+            && !(t1.contains_opaque() || t2.contains_opaque())
         {
             return Ok(());
         }
 
+        if allow_cast && (t1.contains_opaque() || t2.contains_opaque()) {
+            return Err(Error::ExpectOnOpaqueType { location });
+        }
```

Comment:
Note: maybe the check is _too high_ in the function body though 🤔, as we might still want to allow var and links to collapse before raising this error.